### PR TITLE
Make Using lazy

### PR DESCRIPTION
**Pull request by NthPortal**
Make Using lazy, allowing `map` and `filter` to be
called before the resource is released.

### CODE REVIEW

**Comment by NthPortal** [comment]
Bikeshedding welcome on better names for `class Using` and its `use()` method.

Possible other names for `class Using` (i.e. the return type of `Using.apply`):

- `Usage` (h/t @dwijnand)
- `ResourceUse`

Possible other names for `use()`:
- `run()`
- `consume()` (h/t @dwijnand)

File: src/library/scala/util/Using.scala [hunk 1]
```diff
@@ -40,66 +43,85 @@ import scala.util.control.{ControlThrowable, NonFatal}
...
+    */
+  def flatMap[R1: Resource, B](f: A => Using[R1, B]): Using[R, B] =
+    new Using(resource, r => f(op(r)).use0())
+
+  def filter(p: A => Boolean): Using[R, A] =
+    new Using(resource, r => {
+      val res = op(r)
+      if (p(res)) res
+      else throw new NoSuchElementException("Predicate does not hold for " + res)
+    })
```
Comment (julienrf) [inline src/library/scala/util/Using.scala]: This is not really related to this PR, but is it a good idea to have filter if it throws an exception? I think we should not have it.

**Comment by dwijnand** [comment]
👨‍🎨 Maybe scala-arm's `ManagedResource`? 🤔

File: src/library/scala/util/Using.scala (hunk 1 above)
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: It seems relatively related to this PR to me.

The question is, if not `NoSuchElementException`, what does the eventual `Try` contain?

(Side note: I tried out an optimisation where it uses a cached `ControlThrowable` initially, and then only turns it into a `NoSuchElementException` if you ask for a `Try` back. If you call `foreach`, it just catches the cached throwable and does nothing. Other than being a bit confusing from a maintenance perspective, I didn't see any downsides.)

Comment (julienrf) [reply src/library/scala/util/Using.scala]: > The question is, if not `NoSuchElementException`, what does the eventual `Try` contain?

Any exception that can be thrown here: https://github.com/scala/scala/blob/56c45cd8d65e13af688892fc3831bbba9b8a8a28/src/library/scala/util/Using.scala#L181-L217

**Comment by He-Pin** [comment]
@NthPortal useAndRelease？

_使用 [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github) 从我的 Galaxy Note8 发送_

File: src/library/scala/util/Using.scala (hunk 1 above)
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: I'm fine with removing `filter`. If we decide it's a good idea later, it can always be added back.

**Comment by NthPortal** [comment]
@dwijnand I'm happy with `ManagedResource` as long as @jsuereth is okay with me plagiarising the name ;)

**Comment by NthPortal** [comment]
While we're repainting the shed, thoughts on renaming `Using.Resource` to `Using.Release` or `Using.Releaser`?

**Comment by som-snytt** [comment]
If you called it `Usage`, you could use Josh's line from `arm`:

`For more information on usage, see Usage.`

**Comment by Ichoran** [comment]
I always prefer the loan pattern to programmer-implements-linear-types-in-their-head.  So I'm not a fan.

(Loan pattern would look like

```
Using.Manager{ use =>
  for {
    a <- use(foo)
    b <- use(bar)
    c <- use(baz)
  } yield bippy(a, b, quux(c))
}
```

You can screw this up too by assigning intermediates to some mutable value, but it isn't nearly so error-prone for people who aren't used to lazy evaluation.

**Comment by NthPortal** [comment]
@Ichoran in your example, all the resources get released at the `Using.Manager` block, correct? similar to the `Scope` described by @densh in [this Scala World talk](https://www.youtube.com/watch?v=MV2eJkwarT4) ([gist](https://gist.github.com/densh/75d2d3063571d89ee34e161b4a61c74a))?

I'm not opposed to that, though I'm curious as to others' thoughts on whether this is slowly exceeding the scope of what should be in the standard library. Originally, this was just meant to be a simple `Try` wrapper around `Using.resource`.

**Comment by NthPortal** [comment]
@Ichoran once it's managed externally, is there any reason for `use(foo)` to return a monad-ish thing and not just return the resource?

**Comment by Ichoran** [comment]
@NthPortal - Without coding it up I can't be sure, but I don't think the monadish thing is needed any longer.

**Comment by NthPortal** [comment]
Superseded by #7468
